### PR TITLE
core: move tee_ta_manager.c out of arch/arm

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Moves most parts of tee_ta_manager.c into core/kernel, keeping only
tee_ta_verify_param() in the original file as it uses architecture
specific checks.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/kernel/tee_ta_manager.c [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+#include <tee/tee_obj.h>
+#include <tee/tee_svc_storage.h>
+#include <tee_api_types.h>
+#include <trace.h>
+#include <utee_types.h>
+#include <util.h>
+#include <assert.h>
+
+/* This mutex protects the critical section in tee_ta_init_session */
+static struct mutex tee_ta_mutex = MUTEX_INITIALIZER;
```
Comment (pascal-brand-st-dev) [inline core/kernel/tee_ta_manager.c]: Mutexes have arm-specific implementation. Do you think it could be made generic too? may be in another PR?

File: core/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+static struct condvar tee_ta_cv = CONDVAR_INITIALIZER;
+static int tee_ta_single_instance_thread = THREAD_ID_INVALID;
+static size_t tee_ta_single_instance_count;
+struct tee_ta_ctx_head tee_ctxes = TAILQ_HEAD_INITIALIZER(tee_ctxes);
+
+static void lock_single_instance(void)
+{
+	/* Requires tee_ta_mutex to be held */
+	if (tee_ta_single_instance_thread != thread_get_id()) {
```
Comment (pascal-brand-st-dev) [inline core/kernel/tee_ta_manager.c]: thread implementation is ARM specific. This one looks difficult to make entirely generic, because of thread_a32.S.

File: core/kernel/tee_ta_manager.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/kernel/tee_ta_manager.c]: We're already using mutexes under `core/tee` for instance.

I think it should be a requirement of an architecture to provide a `<kernel/mutex.h>` that has all the needed primitives. 
We could of course add a porting layer for the implementation behind `<kernel/mutex.h>`, but without another architecture there's a big risk that we'll miss the target.

File: core/kernel/tee_ta_manager.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/kernel/tee_ta_manager.c]: Yes, but it looks like only `thread_get_id()` need to be provided by an alternative implementation.

File: core/kernel/tee_ta_manager.c (hunk 1 above)
Comment (pascal-brand-st-dev) [inline core/kernel/tee_ta_manager.c]: ok
We definitely miss another target to correctly achieve generic vs specific implementation

File: core/kernel/tee_ta_manager.c (hunk 2 above)
Comment (pascal-brand-st-dev) [inline core/kernel/tee_ta_manager.c]: ok

**Comment by pascal-brand-st-dev** [comment]
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tags applied
